pkg/gitreceive/build.go

```diff
@@ -0,0 +1,295 @@
+}
+
+// run prints the command it will execute to the debug log, then runs it and returns the result of run
+func run(cmd *exec.Cmd) error {
+	cmdStr := strings.Join(cmd.Args, " ")
+	if cmd.Dir != "" {
+		log.Debug("running [%s] in directory %s", cmdStr, cmd.Dir)
+	} else {
+		log.Debug("running [%s]", cmdStr)
+	}
+	return cmd.Run()
+}
+
+func build(conf *Config, builderKey, gitSha string) error {
+	storage, err := getStorageConfig()
+	if err != nil {
+		return err
+	}
+	creds, err := getStorageCreds()
+	if err == errMissingKey || err == errMissingSecret {
+		return err
+	}
+
+	repo := conf.Repository
+	if len(gitSha) <= shortShaIdx {
+		return errGitShaTooShort{sha: gitSha}
+	}
+	shortSha := gitSha[0:8]
+	appName := conf.App()
+
+	repoDir := filepath.Join(conf.GitHome, repo)
+	buildDir := filepath.Join(repoDir, "build")
+
+	slugName := fmt.Sprintf("%s:git-%s", appName, shortSha)
+	imageName := strings.Replace(slugName, ":", "-", -1)
+	if err := os.MkdirAll(buildDir, os.ModeDir); err != nil {
+		return fmt.Errorf("making the build directory %s (%s)", buildDir, err)
+	}
+	tmpDir := os.TempDir()
+
+	tarURL := fmt.Sprintf("%s://%s:%s/git/home/%s/tar", storage.schema(), storage.host(), storage.port(), slugName)
+
+	// this is where workflow tells slugrunner to download the slug from, so we have to tell slugbuilder to upload it to here
+	pushURL := fmt.Sprintf("%s://%s:%s/git/home/%s/push", storage.schema(), storage.host(), storage.port(), fmt.Sprintf("%s:git-%s", appName, gitSha))
+
+	// Ensure that the app config can be gotten from workflow. We don't do anything with this information
+	appConf, err := getAppConfig(conf, builderKey, conf.Username, appName)
+	if err != nil {
+		return fmt.Errorf("getting app config for %s (%s)", appName, err)
+	}
+	log.Debug("got the following config back for app %s: %+v", appName, *appConf)
+	var buildPackURL string
+	if buildPackURLInterface, ok := appConf.Values["BUILDPACK_URL"]; ok {
+		if bpStr, ok := buildPackURLInterface.(string); ok {
+			log.Debug("found custom buildpack URL %s", bpStr)
+			buildPackURL = bpStr
+		}
+	}
+
+	// build a tarball from the new objects
+	gitArchiveCmd := repoCmd(repoDir, "git", "archive", "--format=tar.gz", fmt.Sprintf("--output=%s.tar.gz", appName), gitSha)
+	gitArchiveCmd.Stdout = os.Stdout
+	gitArchiveCmd.Stderr = os.Stderr
+	if err := run(gitArchiveCmd); err != nil {
+		return fmt.Errorf("running %s (%s)", strings.Join(gitArchiveCmd.Args, " "), err)
+	}
+
+	// untar the archive into the temp dir
+	tarCmd := repoCmd(repoDir, "tar", "-xzf", fmt.Sprintf("%s.tar.gz", appName), "-C", fmt.Sprintf("%s/", tmpDir))
+	tarCmd.Stdout = os.Stdout
+	tarCmd.Stderr = os.Stderr
+	if err := run(tarCmd); err != nil {
+		return fmt.Errorf("running %s (%s)", strings.Join(tarCmd.Args, " "), err)
+	}
+
+	usingDockerfile := true
+	rawProcFile, err := ioutil.ReadFile(fmt.Sprintf("%s/Procfile", tmpDir))
+	if err == nil {
+		usingDockerfile = false
```

Comment:
what about apps that have both a Dockerfile and a Procfile? In v1 we assumed that this app is still using a Dockerfile, but the Procfile would be used as the source of truth for process types. It's weird, I know...